kong/clustering/data_plane.lua

```diff
@@ -219,6 +220,17 @@ function _M:communicate(premature)
 
   local conf = self.conf
 
+  do
+    local response_data, err = version_negotiation.request_version_handshake(conf, self.cert, self.cert_key)
+    if not response_data then
+      ngx_log(ngx_ERR, _log_prefix, "error while requesting version negotiation: " .. err)
+      assert(ngx.timer.at(math.random(5, 10), function(premature)
+        self:communicate(premature)
+      end))
+      return
+    end
```

Comment:
on the DP side, the result is stored in a shared dict and available via `get_negotiated_service()`.

Reply:
What I meant was if the negotiation failed (CP doesn't have the necessary services), should this code not error out somewhere (and subsequently retry or give up)?

Reply:
ah, you mean if the CP responds with 200 but a "critical" service is rejected?  I think in that case whatever code requires that service will query the result (with `get_negotiated_service()`) and if it was rejected then it should do whatever it can (probably complain noisily and abort).

I consider this module is strictly a communication mechanism, not policy.  It's up to the users of this feature to establish relevant behaviours.

Reply:
Fair enough, thanks for explaining.